fl4health/losses/deep_mmd_loss.py

```diff
@@ -0,0 +1,224 @@
+from typing import Optional, Tuple
+
+import numpy as np
+import torch
+
+
+class ModelLatentF(torch.nn.Module):
+    """Latent space for both domains."""
+
+    def __init__(self, x_in: int, H: int, x_out: int):
+        """Init latent features."""
+        super(ModelLatentF, self).__init__()
+        self.restored = False
+        self.latent = torch.nn.Sequential(
+            torch.nn.Linear(x_in, H, bias=True),
+            torch.nn.Softplus(),
+            torch.nn.Linear(H, H, bias=True),
+            torch.nn.Softplus(),
+            torch.nn.Linear(H, H, bias=True),
+            torch.nn.Softplus(),
+            torch.nn.Linear(H, x_out, bias=True),
+        )
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        """Forward the LeNet."""
+        fealant = self.latent(input)
+        return fealant
+
+
+class DeepMmdLoss(torch.nn.Module):
+    def __init__(
+        self,
+        device: torch.device,
+        input_size: int,
+        lr: float = 0.001,
+        hidden_size: int = 10,
+        output_size: int = 50,
+        layer_name: Optional[str] = None,
+        training: bool = True,
+    ) -> None:
+        super().__init__()
+        self.device = device
+        self.lr = lr
+        self.layer_name = layer_name
+        self.training = training
+
+        # Initialize the model
+        self.featurizer = ModelLatentF(input_size, hidden_size, output_size).to(self.device)
+
+        # Initialize parameters
+        self.epsilonOPT: torch.Tensor = torch.log(torch.from_numpy(np.random.rand(1) * 10 ** (-10)).to(self.device))
+        self.epsilonOPT.requires_grad = self.training
+        self.sigmaOPT: torch.Tensor = torch.sqrt(torch.tensor(2 * 32 * 32, dtype=torch.float).to(self.device))
+        self.sigmaOPT.requires_grad = self.training
+        self.sigma0OPT: torch.Tensor = torch.sqrt(torch.tensor(0.005, dtype=torch.float).to(self.device))
+        self.sigma0OPT.requires_grad = self.training
+
+        # Initialize optimizers
+        self.optimizer_F = torch.optim.Adam(
+            list(self.featurizer.parameters()) + [self.epsilonOPT] + [self.sigmaOPT] + [self.sigma0OPT], lr=self.lr
+        )
+
+    def Pdist2(self, x: torch.Tensor, y: Optional[torch.Tensor]) -> torch.Tensor:
```

Comment:
maybe expand this to `pairwise_distiance_squared`?

Reply:
It looks like we don't leverage the fact that y can be none to get the distances of x with itself. Maybe we just drop that option and require y to be passed to simplify this function.